telegram/ext/_fsm/machine.py

```diff
@@ -0,0 +1,200 @@
+        )
+        self._states = MappingProxyType(self.__states)
+
+        self.__job_queue: Optional[weakref.ReferenceType[JobQueue]] = None
+        self.__history: Mapping[_KT, MutableSequence[State]] = defaultdict(
+            lambda: deque(maxlen=10)
+        )
+
+    @property
+    def states(self) -> Mapping[_KT, tuple[State, int]]:
+        return self._states
+
+    def store_state_history(self, key: _KT, state: State) -> None:
+        # Making this public so that users can override if they want to customize the history
+        # E.g., they could want to store more/fewer states, also depending on the key
+        self.__history[key].append(state)
+
+    def get_state_history(self, key: _KT) -> Sequence[State]:
+        return list(self.__history[key])
+
+    def get_lock(self, key: _KT) -> asyncio.Lock:
+        """Returns a lock that is unique for this key at runtime.
+        It can be used to prevent concurrent access to resources associated to this key.
+        """
+        return self._locks.setdefault(key, asyncio.Lock())
+
+    @abc.abstractmethod
+    def get_state_info(self, update: object) -> StateInfo[_KT]:
+        """Returns exactly one active state for the update.
+        If more than one stored key applies to the update, one must be chosen.
+        It's recommended to select the most specific one.
+
+        Example:
+            The state of a chat, a user or a user in a specific chat could be tracked.
+            For a message in that chat, the state of the user in that chat should be returned if
+            available. Otherwise, the state of the chat should be returned.
+
+        Important:
+            This must be an atomic operation and not e.g. wait for a lock.
+            Instead, if necessary, return a special state indicating that the key is currently
+            busy.
+        """
+
+    def _do_set_state(
+        self, key: _KT, state: State, version: Optional[int] = None
+    ) -> StateInfo[_KT]:
+        """Protected method to set the state for the specified key.
+
+        The version can be optionally used for optimistic locking. If the version does not match
+        the current version, the state should not be updated.
+
+        Important:
+            This should be used exclusively by methods of this class and subclasses.
+            It should *not* be called directly by users of this class!
+        """
+        _LOGGER.debug("Setting %s state to %s", key, state)
+        if state is State.ANY:
+            raise ValueError("State.ANY is not supported in set_state")
+
+        if version and version != self._states.get(key, (None, None))[1]:
+            raise ValueError("Optimistic locking failed. Not updating state.")
+
+        if jq := self._get_job_queue(raise_exception=False):
+            # This is a rather tight coupling between FSM and JobQueue
+            # Not sure if we like that. Makes it even harder to replace JobQueue
```

Comment:
No I think it is fine. We can always do the work when for some reason we have the need to but for now we should enjoy it imo